From dd10d1adef57627d84ce1cf8c3ba99156fbccb6c Mon Sep 17 00:00:00 2001
From: Vivek Goyal <vgoyal@redhat.com>
Date: Tue, 19 Aug 2014 22:02:53 -0400
Subject: [security] keys: special dot prefixed keyring name bug fix

Message-id: <1408485785-777-22-git-send-email-vgoyal@redhat.com>
Patchwork-id: 88384
O-Subject: [RHEL7.1 PATCH 21/33] KEYS: special dot prefixed keyring name bug fix
Bugzilla: 1129779
RH-Acked-by: David Howells <dhowells@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

BZ:1129779

commit a4e3b8d79a5c6d40f4a9703abf7fe3abcc6c3b8d
Author: Mimi Zohar <zohar@linux.vnet.ibm.com>
Date: Thu, 22 May 2014 14:02:23 -0400

Dot prefixed keyring names are supposed to be reserved for the
kernel, but add_key() calls key_get_type_from_user(), which
incorrectly verifies the 'type' field, not the 'description' field.
This patch verifies the 'description' field isn't dot prefixed,
when creating a new keyring, and removes the dot prefix test in
key_get_type_from_user().

Changelog v6:
- whitespace and other cleanup

Changelog v5:
- Only prevent userspace from creating a dot prefixed keyring, not
  regular keys  - Dmitry

Reported-by: Dmitry Kasatkin <d.kasatkin@samsung.com>
Signed-off-by: Mimi Zohar <zohar@linux.vnet.ibm.com>
Acked-by: David Howells <dhowells@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/security/keys/keyctl.c b/security/keys/keyctl.c
index 7e791bc..da50c93 100644
--- a/security/keys/keyctl.c
+++ b/security/keys/keyctl.c
@@ -37,8 +37,6 @@ static int key_get_type_from_user(char *type,
   return ret;
  if (ret == 0 || ret >= len)
   return -EINVAL;
- if (type[0] == '.')
-  return -EPERM;
  type[len - 1] = '\0';
  return 0;
 }
@@ -86,6 +84,10 @@ SYSCALL_DEFINE5(add_key, const char __user *, _type,
   if (!*description) {
    kfree(description);
    description = NULL;
+  } else if ((description[0] == '.') &&
+      (strncmp(type, "keyring", 7) == 0)) {
+   ret = -EPERM;
+   goto error2;
   }
  }
 
-- 
1.7.1